main.nf

```diff
@@ -92,44 +53,671 @@ if (workflow.profile.contains('awsbatch')) {
+}
+
+// Pipeline OS-specific commands
+ZCAT = (System.properties['os.name'] == 'Mac OS X' ? 'gzcat' : 'zcat')
+
+
+/**
+ * Download data from ensembl for the particular species.
  */
-if (params.input_paths) {
-    if (params.single_end) {
-        Channel
-            .from(params.input_paths)
-            .map { row -> [ row[0], [ file(row[1][0], checkIfExists: true) ] ] }
-            .ifEmpty { exit 1, "params.input_paths was empty - no input files supplied" }
-            .into { ch_read_files_fastqc; ch_read_files_trimming }
-    } else {
-        Channel
-            .from(params.input_paths)
-            .map { row -> [ row[0], [ file(row[1][0], checkIfExists: true), file(row[1][1], checkIfExists: true) ] ] }
-            .ifEmpty { exit 1, "params.input_paths was empty - no input files supplied" }
-            .into { ch_read_files_fastqc; ch_read_files_trimming }
-    }
-} else {
-    Channel
-        .fromFilePairs(params.input, size: params.single_end ? 1 : 2)
-        .ifEmpty { exit 1, "Cannot find any reads matching: ${params.input}\nNB: Path needs to be enclosed in quotes!\nIf this is single-end data, please specify --single_end on the command line." }
-        .into { ch_read_files_fastqc; ch_read_files_trimming }
+process ensembl_fasta_download{
+
+   when:
+   params.add_reference ||  params.ensembl || params.altorfs || params.ncrna || params.pseudogenes || params.vcf
+
+   input:
+   file ensembl_downloader_config
+
+   output:
+   file 'database_ensembl/*.pep.all.fa' into ensembl_protein_database_sub
+   file 'database_ensembl/*cdna.all.fa' into ensembl_cdna_database, ensembl_cdna_database_sub
+   file 'database_ensembl/*ncrna.fa' into ensembl_ncrna_database, ensembl_ncrna_database_sub
+   file 'database_ensembl/*.dna*.fa' into genome_fasta
+   file 'database_ensembl/*.gtf' into gtf
+
+   script:
+   """
+   pypgatk_cli.py ensembl-downloader --config_file ${ensembl_downloader_config} --ensembl_name ${params.ensembl_name} \\
+                  -sv -sc
+   """
+}
+
+process add_reference_proteome{
+
+   when:
+   params.add_reference
+
+   input:
+   file reference_proteome from ensembl_protein_database_sub
+
+   output:
+   file 'reference_proteome.fa' into ensembl_protein_database
+
+   script:
+   """
+   cat ${reference_proteome} >> reference_proteome.fa
+   """
+
+}
+
+/**
+ * Concatenate cDNA and ncRNA databases
+ **/
+process merge_cdnas{
+
+   input:
+   file a from ensembl_cdna_database_sub.collect()
+   file b from ensembl_ncrna_database_sub.collect()
+
+   output:
+   file 'total_cdnas.fa' into total_cdnas
+
+   script:
+   """
+   cat ${a} >> total_cdnas.fa
+   cat ${b} >> total_cdnas.fa
+   """
+}
+
+/**
+ * Creates the ncRNA protein database
+ */
+process add_ncrna{
+
+   publishDir "${params.outdir}", mode: 'copy', overwrite: true
```

Comment:
You said that only a single file was produced by this pipeline, right? But in every process here you have `publishDir` set up to save all of the `output` files.

If this isn't intentional and these are intermediate files only that are not worth keeping by the end user, then we should remove all of the `publishDir` lines, as they are just doubling all of the pipeline files on the disk for no reason.

Reply:
I have removed all the intermedia files

Reply:
ok now you removed _all_ of the `publishDir` statements, which means that the pipeline doesn't generate any output files at all 😅  There's probably a middle ground here somewhere..!

Reply:
Refactored in https://github.com/nf-core/pgdb/pull/35